src/main/java/filter/expression/QualifierType.java

```diff
@@ -0,0 +1,78 @@
+
+package filter.expression;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+public enum QualifierType {
+
+    EMPTY, // The empty qualifier, which always returns true
+    FALSE, // The false qualifier, which always returns false
+
+    KEYWORD, // A special value that indicates a search term
+
+    ASSIGNEE, AUTHOR, CREATED, DATE, DESCRIPTION, HAS, ID, IN,
+    INVOLVES, IS, LABEL, LABELS, MILESTONE, MILESTONES, NO,
+    REPO, SORT, STATE, TITLE, TYPE, UPDATED;
+
+    private static final Map<String, QualifierType> ALIASES = initialiseAliases();
+
+    public static Map<String, QualifierType> initialiseAliases() {
+        Map<String, QualifierType> aliases = new HashMap<>();
+        aliases.put("m", MILESTONE);
+        aliases.put("as", ASSIGNEE);
+        aliases.put("au", AUTHOR);
+        aliases.put("s", STATE);
+        aliases.put("creator", AUTHOR);
+        aliases.put("user", INVOLVES);
+        aliases.put("status", STATE);
+        aliases.put("body", DESCRIPTION);
+        aliases.put("desc", DESCRIPTION);
+        return Collections.unmodifiableMap(aliases);
+    }
+
+    /**
+     * Returns a list of keywords which may be completed when a prefix is typed in
+     * the filter field. Defaults to the string forms of all qualifier types. Exceptions
+     * may be added by modifying the collection before it is returned, for example, to
+     * include unique aliases or remove conflicting ones ('is' conflicts with 'issue')
+     *
+     * @return
+     */
+    public static Set<String> getCompletionKeywords() {
+        Set<String> defaultCompletions = new ArrayList<>(Arrays.asList(QualifierType.values())).stream()
+            .map(QualifierType::name)
+            .map(String::toLowerCase)
+            .collect(Collectors.toSet());
+
+        defaultCompletions.addAll(Arrays.asList(
+            "creator", "user", "status", "body", "closed", "open", "issue", "pr", "pullrequest",
```

Comment:
Okay, I'll use aliases for the common ones, and `remove` those which shouldn't appear. Filtering by length is a heuristic and works here, but it may not for other cases. Probably better to make exceptions explicitly.

'desc' is a prefix of 'description' (like 'is' and 'issue'), so I left it out. I suppose there's no real harm in including them.